### PR TITLE
concurrent file merging in passthrough mode to reduce initial pull time

### PR DESCRIPTION
**Pull request by wswsmao**
In passthrough mode, the initial pull of an image requires merging chunks into a file. This process can be time-consuming, especially for large files.

This pr optimizes the merging process by introducing two configuration options: `merge_buffer_size` and `merge_worker_count`. By concurrently reading chunks and caching them for batch writing, we enhance the performance of the initial image pull in passthrough mode.

The principle of concurrent reading is illustrated in the diagram below. 
The `merge_worker_count` corresponds to the number of workers in the diagram, while `merge_buffer_size` corresponds to the size of the buffer in the diagram. Before concurrent reading begins, the buffer area is divided, and each worker is assigned a specific region. This approach allows all workers to concurrently read the contents of the chunks into the buffer.
image

### CODE REVIEW

**Comment by wswsmao** [comment]
cc @ktock

File: fs/reader/reader.go [hunk 1]
```diff
@@ -542,78 +555,122 @@ func (sf *file) GetPassthroughFd() (uintptr, error) {
...
+		for err := range errChan {
+			if err != nil {
+				w.Abort()
+				return err
+			}
 		}
-		if _, err := w.Write(ip); err != nil {
-			sf.gr.putBuffer(b)
+
+		if _, err := w.Write(buffer); err != nil {
```
Comment (ktock) [inline fs/reader/reader.go:635]: Needs to guarantee, check and test exhaustiveness

File: fs/reader/reader.go [hunk 2]
```diff
@@ -542,78 +555,122 @@ func (sf *file) GetPassthroughFd() (uintptr, error) {
...
-		// cache miss, prefetch the whole chunk
-		if _, err := sf.fr.ReadAt(ip, chunkOffset); err != nil && err != io.EOF {
-			sf.gr.putBuffer(b)
-			w.Abort()
-			return fmt.Errorf("failed to read data: %w", err)
+		batchSize := batchEnd - batchStart
+		buffer := make([]byte, batchSize)
+
+		var wg sync.WaitGroup
+		errChan := make(chan error, len(batchChunks))
```
Comment (ktock) [inline fs/reader/reader.go:610]: use errgroup

File: fs/reader/reader.go [hunk 3]
```diff
@@ -542,78 +555,122 @@ func (sf *file) GetPassthroughFd() (uintptr, error) {
...
+			batchChunks = append(batchChunks, chunks[i])
 		}
 
-		// cache miss, prefetch the whole chunk
-		if _, err := sf.fr.ReadAt(ip, chunkOffset); err != nil && err != io.EOF {
-			sf.gr.putBuffer(b)
-			w.Abort()
-			return fmt.Errorf("failed to read data: %w", err)
+		batchSize := batchEnd - batchStart
+		buffer := make([]byte, batchSize)
```
Comment (ktock) [inline fs/reader/reader.go:607]: Needs mutex for all shared variables

File: docs/passthrough.md [hunk 4]
```diff
@@ -38,6 +40,14 @@ After updating the configuration, specify the `config.toml` file when starting `
 $ containerd-stargz-grpc -config config.toml
 ```
 
+## Advanced Configuration
+
+In passthrough mode, the initial pull of an image requires merging chunks into a file. This process can be time-consuming, especially for large files.
+
+To optimize the time taken for the initial image pull, you can use the `merge_buffer_size` and `merge_worker_count` configuration options. The `merge_buffer_size` specifies the size of the buffer used for reading the image, with a default value of 400MB. The `merge_worker_count` determines the level of concurrency for reading the image, with a default value of 10.
+
+By concurrently reading chunks and caching them for batch writing, you can significantly enhance the performance of the initial image pull in passthrough mode.
```
Comment (ktock) [inline docs/passthrough.md:49]: > significantly enhance the performance 

Do you have benchmarks?

Comment (wswsmao) [reply docs/passthrough.md:49]: I have pushed the test image to docker hub: `abushwang/cmerge:514M-c7717cb6`. 

The image information is as follows, where the `sha512sum` command is executed on a large file. The value of the large file is intentionally chosen to avoid being aligned to 512 bytes, in order to cover scenarios that test the boundaries of the buffer.

```
nerdctl inspect abushwang/cmerge:514M-c7717cb6
...
            "Cmd": [
                "sh",
                "-c",
                "sha512sum /data/514MB"
            ],
...
```

1. Execute `sha512sum` locally to record the original SHA value of the file:
```
sha512sum 514MB
c7717cb60f2b1400def55bb3451ab58004d7c6c8ecc2c0982a9828d86eb7826d730f711a01c9af982f447a0958b50933dbba4ec5d7f8f33de0b299385c19c94c  514M
```

2. Compare the time taken before and after optimization:
**Before:**
```
time nerdctl --insecure-registry run --rm -it --snapshotter stargz $images_name 
...

9.30.17.238:5000/ocs9:514-sha512-esgz:                                            resolved       |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:ba9f6fcca13ffcbc05bfbb0a7197d1ef9541e45fcfcb50cb778a3c4930240c70: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:42d6f47a2f5eea7e5b65730666273aa258999886989841edac42d9023bf3c77d:   done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 0.1 s                                                                    total:   0.0 B (0.0 B/s)                                         
c7717cb60f2b1400def55bb3451ab58004d7c6c8ecc2c0982a9828d86eb7826d730f711a01c9af982f447a0958b50933dbba4ec5d7f8f33de0b299385c19c94c  /data/514MB

real    0m8.584s
user    0m0.274s
sys     0m0.068s
```

**After:**
```
time nerdctl --insecure-registry run --rm -it --snapshotter stargz $images_name 
...                                      
9.30.17.238:5000/ocs9:514-sha512-esgz:                                            resolved       |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:ba9f6fcca13ffcbc05bfbb0a7197d1ef9541e45fcfcb50cb778a3c4930240c70: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:42d6f47a2f5eea7e5b65730666273aa258999886989841edac42d9023bf3c77d:   done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 0.1 s                                                                    total:   0.0 B (0.0 B/s)                                         
c7717cb60f2b1400def55bb3451ab58004d7c6c8ecc2c0982a9828d86eb7826d730f711a01c9af982f447a0958b50933dbba4ec5d7f8f33de0b299385c19c94c  /data/514MB

real    0m4.891s
user    0m0.288s
sys     0m0.037s
```

From the test results, we can conclude that:
1. The SHA value of the file remains unchanged after optimization, ensuring the integrity of the file.
2. There is a significant performance improvement compared to before optimization.

File: fs/reader/reader.go (hunk 3 above)
Comment (wswsmao) [reply fs/reader/reader.go:607]: It seems that there is no need to lock the buffer. I have added a diagram and explanation in the PR description as follows. 
image

Specifically, before concurrent reading begins, the buffer area is divided, and each worker is assigned a specific region. As a result, the operations of each worker on the buffer do not conflict with one another.

File: fs/reader/reader.go (hunk 2 above)
Comment (wswsmao) [reply fs/reader/reader.go:610]: done

File: fs/reader/reader.go (hunk 1 above)
Comment (wswsmao) [reply fs/reader/reader.go:635]: The operation of verifyOneChunk is within processBatchChunks:
https://github.com/containerd/stargz-snapshotter/pull/1987/files#diff-36a27f22f302d0e12d76075d692f7794795538342d157afb21a5afa59b304b8fR667

File: fs/reader/reader.go (hunk 2 above)
Comment (wswsmao) [reply fs/reader/reader.go]: done

File: fs/reader/reader.go (hunk 1 above)
Comment (ktock) [reply fs/reader/reader.go:627]: At least test should be added for detecting incomplete writes to`buffer` e.g. a portion of the buffer is empty because of wrong index calculation.

**Review by ktock — APPROVED** [review]
Thanks, merging for now. https://github.com/containerd/stargz-snapshotter/pull/1987#discussion_r1988314111 should be done in following ups.

**Comment by wswsmao** [comment]
> Thanks, merging for now. [#1987 (comment)](https://github.com/containerd/stargz-snapshotter/pull/1987#discussion_r1988314111) should be done in following ups.

ok，I will finish it in following pr
